fix(ivy): i18n - support setting locales for each translation file

petebacondarwin opened the pull request:
Previously the target locale of a translation file had to be extracted
from the contents of the translation file.

Now an array of locales can be provided via the `translationFileLocales`
option that overrides any target locale extracted from the file.
This allows us to support translation files that do not have a target
locale specified in their contents.

// FW-1644
Fixes #33323

petebacondarwin committed:
fix(ivy): i18n - support setting locales for each translation file

Previously the target locale of a translation file had to be extracted
from the contents of the translation file. Therefore it was an error if
the translation file did not provide a target locale.

Now an array of locales can be provided via the `translationFileLocales`
option that overrides any target locale extracted from the file.
This allows us to support translation files that do not have a target
locale specified in their contents.

// FW-1644
Fixes #33323

clydin commented on packages/localize/src/tools/src/translate/translation_files/translation_file_loader.ts:
```diff
@@ -20,12 +20,13 @@ export class TranslationLoader {
       for (const translationParser of this.translationParsers) {
         if (translationParser.canParse(filePath, fileContents)) {
-          return translationParser.parse(filePath, fileContents);
+          return translationParser.parse(filePath, fileContents, translationFileLocales[index]);
```
Can the logic to set the locale if not present in the parsed file be moved up here after it is parsed?
The CLI already has the locale information so it would have to pass it in and then get it right back to avoid the missing locale exception.

This also prevents the CLI from performing a validity check on the parsed file locale since it will now be wiped by the passed in locale.

petebacondarwin replied:
Ok

IgorMinar commented on packages/localize/src/tools/src/translate/translation_files/translation_parsers/translation_parser.ts:
```diff
@@ -25,5 +25,5 @@ export interface TranslationParser {
    * @param contents The contents of the translation file.
    */
-  parse(filePath: string, contents: string): TranslationBundle;
+  parse(filePath: string, contents: string, locale: string|undefined): TranslationBundle;
```
why not this? 
```suggestion
  parse(filePath: string, contents: string, locale?: string): TranslationBundle;
```

petebacondarwin replied:
They look them same but have slightly different semantics. 
In the case you are suggesting the argument is optional.
In the case in the PR the argument is required but can be undefined.
I like making arguments required if possible because t helps identify places that are forgetting to make a decision about the argument. Also in this case it really is the case that you might pass undefined rather than not specifying the argument.

As it turns out the request from @clydin will probably mean that a lot of this will go away anyway.

IgorMinar replied:
I see your point, and because we are still falling back onto the locale defined in the translation file, it would be helpful to be explicit. In that case I suggest that you document this behavior in the jsdoc above.

IgorMinar commented on packages/localize/src/tools/src/translate/translation_files/translation_parsers/xliff1/xliff1_translation_parser.ts:
```diff
@@ -30,30 +30,34 @@ export class Xliff1TranslationParser implements TranslationParser {
   }
 
-  parse(filePath: string, contents: string): TranslationBundle {
+  parse(filePath: string, contents: string, locale: string|undefined): TranslationBundle {
```
same as above

IgorMinar commented on packages/localize/src/tools/test/translate/translation_files/translation_parsers/simple_json/simple_json_spec.ts:
```diff
@@ -20,18 +20,27 @@ describe('SimpleJsonTranslationParser', () => {
       const parser = new SimpleJsonTranslationParser();
-      const result = parser.parse('/some/file.json', '{"locale": "en", "translations": {}}');
+      const result =
+          parser.parse('/some/file.json', '{"locale": "en", "translations": {}}', undefined);
```
you wouldn't need this if you made it optional

IgorMinar reviewed (approved):
the rest lgtm

petebacondarwin committed:
fix(ivy): i18n - update `localize-translate` to accept target-locales

The `localize-translate` command line tool can now accept an array of
target locales to support the case where translation files do not
contain them. Specify this array via the `--target-locales` option.

NOTE to early adopters: in order to support this, the original `-t`
option for the binary has changed from being a glob pattern to an array
of paths, which will have matching indices to any provided target-locales.

petebacondarwin commented:
After @clydin's feedback I completely redesigned this PR - so all the files have changed since the last reviews (cc @IgorMinar).

@clydin please note that the return type from `TranslationParser.parse()` has changed. I think you are using this directly in the CLI integration?? So this will be a breaking change for this integration.

PTAL!!

IgorMinar commented on packages/localize/src/tools/src/translate/translation_files/translation_loader.ts:
```diff
@@ -0,0 +1,57 @@
+          if (parsedLocale !== undefined && providedLocale !== undefined &&
+              parsedLocale !== providedLocale) {
+            this.diagnostics.warn(
+                `The provided locale "${providedLocale}" and the target locale "${parsedLocale}" do not match for translation file "${filePath}".`);
```
I don't think that this is easy to understand. How about "specified locale XX doesn't match locale YY found in the translation file"

IgorMinar commented on packages/localize/src/tools/test/translate/translation_files/translation_loader_spec.ts:
```diff
@@ -49,30 +52,71 @@ describe('TranslationLoader', () => {
+      const loader = new TranslationLoader([parser], diagnostics);
+      expect(() => loader.loadBundles(['/src/locale/messages.en.xlf'], []))
+          .toThrowError(
+              'No target locale provided nor found in the translation file "/src/locale/messages.en.xlf"');
```
Maybe add "please provide --target-locale argument"?

petebacondarwin replied:
The only problem with this is that it might be called via the Angular CLI which does not have this option...

petebacondarwin replied:
How about: "consider explicitly providing a target-locale for this file"?

IgorMinar reviewed (commented):
A few more suggestions.

clydin commented on packages/localize/src/tools/src/translate/translation_files/translation_parsers/translation_parser.ts:
```diff
@@ -5,7 +5,15 @@
+*/
+export interface ParsedTranslationBundle {
+  parsedLocale: string|undefined;
+  parsedTranslations: Record<ɵMessageId, ɵParsedTranslation>;
```
If these names are kept unchanged, no code changes would be needed on the CLI side.

petebacondarwin replied:
This is where it is used in the CLI (for context): https://github.com/angular/angular-cli/blob/395c97912800af435021432e708cbdfab28783fe/packages/angular_devkit/build_angular/src/utils/load-translations.ts

Only the `translations` property appears to be important...

petebacondarwin replied:
I'll switch them back...

petebacondarwin commented:
As I am a code-owner of this, and having approval from @IgorMinar, I think that this can be merged, right?

AndrewKushnir commented on packages/localize/src/tools/test/translate/translation_files/translation_parsers/simple_json/simple_json_spec.ts:
```diff
@@ -20,7 +20,7 @@ describe('SimpleJsonTranslationParser', () => {
     it('should extract the locale from the JSON contents', () => {
       const parser = new SimpleJsonTranslationParser();
-      const result = parser.parse('/some/file.json', '{"locale": "en", "translations": {}}');
+      const result = parser.parse('/some/file.json', '{"locale": "en", "translations": {}}', );
```
Extra comma?

AndrewKushnir commented on packages/localize/src/tools/test/translate/translation_files/translation_parsers/simple_json/simple_json_spec.ts:
```diff
@@ -31,7 +31,7 @@ describe('SimpleJsonTranslationParser', () => {
           "Hello, {$ph_1}!": "Bonjour, {$ph_1}!"
         }
-      }`);
+      }`, );
```
Extra comma?

AndrewKushnir commented on packages/localize/src/tools/src/translate/main.ts:
```diff
@@ -47,58 +47,103 @@ if (require.main === module) {
 }
 
 export interface TranslateFilesOptions {
+  /**
```
Thanks for adding docs 👍

AndrewKushnir reviewed (approved):
LGTM, just left a couple comments. Thanks Pete.
